File: backend/src/app/schemas/custom_data.py
"""Custom data API schemas."""

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.common import ORMBaseSchema

ALLOWED_FIELD_TYPES = {"TEXT", "INTEGER", "BOOLEAN", "DATE", "FLOAT"}


# ── Custom Fields ──────────────────────────────────────────────────────────────


class CustomFieldDefinitionCreate(BaseModel):
    field_name: str = Field(min_length=1, max_length=100, pattern=r"^[a-z_][a-z0-9_]*$")
    field_type: str = Field(min_length=1, max_length=20)
    display_name: str = Field(min_length=1, max_length=255)
    sort_order: int = 0


class CustomFieldDefinitionRead(ORMBaseSchema):
    id: uuid.UUID
    customer_id: uuid.UUID
    field_name: str
    field_type: str
    display_name: str
    sort_order: int
    created_at: datetime
    created_by: uuid.UUID | None


class CustomFieldValuesUpdate(BaseModel):
    """Payload: dict of field_name → value."""

    values: dict[str, Any]


class CustomFieldValuesRead(BaseModel):
    values: dict[str, Any]


# ── Custom Tables ──────────────────────────────────────────────────────────────


class CustomColumnCreate(BaseModel):
    column_name: str = Field(min_length=1, max_length=100, pattern=r"^[a-z_][a-z0-9_]*$")
    column_type: str = Field(min_length=1, max_length=20)
    display_name: str = Field(min_length=1, max_length=255)
    sort_order: int = 0


class CustomColumnRead(ORMBaseSchema):
    id: uuid.UUID
    column_name: str
    column_type: str
    display_name: str
    sort_order: int


class CustomTableCreate(BaseModel):
    table_slug: str = Field(min_length=1, max_length=100, pattern=r"^[a-z_][a-z0-9_]*$")
    display_name: str = Field(min_length=1, max_length=255)
    columns: list[CustomColumnCreate] = Field(min_length=1)


class CustomTableRead(ORMBaseSchema):
    id: uuid.UUID
    customer_id: uuid.UUID
    table_slug: str
    display_name: str
    db_table_name: str
    sort_order: int
    created_at: datetime
    created_by: uuid.UUID | None
    columns: list[CustomColumnRead]


# ── Row CRUD ───────────────────────────────────────────────────────────────────


class RowCreate(BaseModel):
    data: dict[str, Any]


class RowUpdate(BaseModel):
    data: dict[str, Any]


class RowRead(BaseModel):
    id: int
    data: dict[str, Any]


class RowsListRead(BaseModel):
    items: list[dict[str, Any]]
    count: int

File: frontend/src/hooks/customData.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { apiClient } from '@/lib/axios'

const BASE = '/api/v1/customers'

// ── Types ─────────────────────────────────────────────────────────────────────

export interface CustomFieldDefinition {
  id: string
  customer_id: string
  field_name: string
  field_type: string
  display_name: string
  sort_order: number
  created_at: string
  created_by: string | null
}

export interface CustomFieldDefinitionCreate {
  field_name: string
  field_type: string
  display_name: string
  sort_order?: number
}

export interface CustomFieldValues {
  values: Record<string, unknown>
}

export interface CustomTableDefinition {
  id: string
  customer_id: string
  table_slug: string
  display_name: string
  db_table_name: string
  sort_order: number
  created_at: string
  created_by: string | null
  columns: CustomColumnDefinition[]
}

export interface CustomColumnDefinition {
  id: string
  column_name: string
  column_type: string
  display_name: string
  sort_order: number
}

export interface CustomTableCreate {
  table_slug: string
  display_name: string
  columns: { column_name: string; column_type: string; display_name: string; sort_order?: number }[]
}

export interface RowsListResponse {
  items: Record<string, unknown>[]
  count: number
}

// ── Custom Fields ─────────────────────────────────────────────────────────────

export function useCustomFieldDefinitions(customerId: string | undefined) {
  return useQuery({
    queryKey: ['custom-fields-defs', customerId],
    queryFn: async () => {
      const { data } = await apiClient.get<CustomFieldDefinition[]>(
        `${BASE}/${customerId}/custom-fields/definitions`
      )
      return data
    },
    enabled: !!customerId,
  })
}

export function useCreateCustomFieldDefinition(customerId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: CustomFieldDefinitionCreate) => {
      const { data } = await apiClient.post<CustomFieldDefinition>(
        `${BASE}/${customerId}/custom-fields/definitions`,
        payload
      )
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['custom-fields-defs', customerId] })
    },
  })
}

export function useDeleteCustomFieldDefinition(customerId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (fieldId: string) => {
      await apiClient.delete(`${BASE}/${customerId}/custom-fields/definitions/${fieldId}`)
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['custom-fields-defs', customerId] })
      qc.invalidateQueries({ queryKey: ['custom-fields-values', customerId] })
    },
  })
}

export function useCustomFieldValues(customerId: string | undefined) {
  return useQuery({
    queryKey: ['custom-fields-values', customerId],
    queryFn: async () => {
      const { data } = await apiClient.get<CustomFieldValues>(
        `${BASE}/${customerId}/custom-fields`
      )
      return data
    },
    enabled: !!customerId,
  })
}

export function useUpdateCustomFieldValues(customerId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (values: Record<string, unknown>) => {
      const { data } = await apiClient.patch<CustomFieldValues>(
        `${BASE}/${customerId}/custom-fields`,
        { values }
      )
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['custom-fields-values', customerId] })
    },
  })
}

// ── Custom Tables ─────────────────────────────────────────────────────────────

export function useCustomTables(customerId: string | undefined) {
  return useQuery({
    queryKey: ['custom-tables', customerId],
    queryFn: async () => {
      const { data } = await apiClient.get<CustomTableDefinition[]>(
        `${BASE}/${customerId}/custom-tables`
      )
      return data
    },
    enabled: !!customerId,
  })
}

export function useCreateCustomTable(customerId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (payload: CustomTableCreate) => {
      const { data } = await apiClient.post<CustomTableDefinition>(
        `${BASE}/${customerId}/custom-tables`,
        payload
      )
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['custom-tables', customerId] })
    },
  })
}

export function useDeleteCustomTable(customerId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (tableId: string) => {
      await apiClient.delete(`${BASE}/${customerId}/custom-tables/${tableId}`)
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['custom-tables', customerId] })
    },
  })
}

export function useTableRows(customerId: string | undefined, tableId: string | undefined) {
  return useQuery({
    queryKey: ['custom-table-rows', customerId, tableId],
    queryFn: async () => {
      const { data } = await apiClient.get<RowsListResponse>(
        `${BASE}/${customerId}/custom-tables/${tableId}/rows`
      )
      return data
    },
    enabled: !!customerId && !!tableId,
  })
}

export function useInsertRow(customerId: string, tableId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (rowData: Record<string, unknown>) => {
      const { data } = await apiClient.post(
        `${BASE}/${customerId}/custom-tables/${tableId}/rows`,
        { data: rowData }
      )
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['custom-table-rows', customerId, tableId] })
    },
  })
}

export function useUpdateRow(customerId: string, tableId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({ rowId, rowData }: { rowId: number; rowData: Record<string, unknown> }) => {
      const { data } = await apiClient.patch(
        `${BASE}/${customerId}/custom-tables/${tableId}/rows/${rowId}`,
        { data: rowData }
      )
      return data
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['custom-table-rows', customerId, tableId] })
    },
  })
}

export function useDeleteRow(customerId: string, tableId: string) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (rowId: number) => {
      await apiClient.delete(
        `${BASE}/${customerId}/custom-tables/${tableId}/rows/${rowId}`
      )
    },
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ['custom-table-rows', customerId, tableId] })
    },
  })
}
